src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java

```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
     }
 
     /**
-     * Uncompresses a single unit of multiple compressed blocks.
+     * Decompresses a single unit of multiple compressed blocks.
      *
      * @param compressed   the compressed data (in.)
      * @param uncompressed the uncompressed data (out.)
+     * @return the actual length of the uncompressed data.
      * @throws IOException if the decompression fails.
      */
-    public static void unCompressUnit(final byte[] compressed,
-                                      final byte[] uncompressed) throws IOException {
+    public int decompressUnit(final byte[] compressed,
+                              final byte[] uncompressed) throws IOException {
 
         // This is just a convenient way to simulate the original code's pointer arithmetic.
         // I tried using buffers but positions in those are always from the beginning and
         // I had to also maintain a position from the start of the current block.
         final OffsetByteArray compressedData = new OffsetByteArray(compressed);
         final OffsetByteArray uncompressedData = new OffsetByteArray(uncompressed);
 
-        for (int i = 0; i * BLOCK_SIZE < uncompressed.length; i++) {
-            final int consumed = uncompressBlock(compressedData, uncompressedData);
+        // "compressedData.offset < compressed.length - 1" means we can read in at least two bytes (the 16-bit header)
+        // TODO,
+        //  figure out why "compressedData.offset < compressed.length" is not the end condition.
+        //  It indicates that somewhere we didn't read the compressed array correctly before reaching here..
+        while (compressedData.offset < compressed.length - 1) {
+            // Bits [11:0] contain the size of the compressed chunk, minus three bytes.
```

Comment:
minus three bytes?  You seem to be adding 3 bytes.  Also why is this required?  Is this code to try and round compressedChunkSize to be a 4 byte multiple or is it really 3 bytes required?